scanners/seo.py

```diff
@@ -0,0 +1,146 @@
+    # Because we're not reading robots.txt, we only need the status code.
+    # We could put these in `pages` but since we have specialized needs for each
+    # and no need to iterate through them, I don't want to.
+    results = {
+        'robots': requests.get("https://" + domain + '/robots.txt').status_code,
+        'warnings': {}
+    }
+    # Now let's take a look at sitemap
+    sitemap = requests.get("https://" + domain + '/sitemap.xml')
+    sitemap_status = sitemap.status_code
+    results['sitemap'] = {
+        'status': sitemap_status
+    }
+    # If we have found a sitemap, see how many URLs are in there?
+    # TO DO: If sitemap.xml redirects to a sitemap index,
+    # we're not handling that correctly yet. We need to detect that
+    # redirect, then ... what? Follow from there?
+    if sitemap_status == 200:
+        soup = BeautifulSoup(sitemap.text, 'xml')
+        urls = soup.find_all('url')
+        results['sitemap']['urls found'] = len(urls)
+        # and how many of those URLs appear to be PDFs
+        if len(urls) > 0:
+            pdfcount = len([u for u in urls if '.pdf' in u.get_text()])
+            results['sitemap']['PDFs found in sitemap'] = pdfcount
+
+    # We'll write to these empty lists for simple dupe checking later
+    titles = []
+    descriptions = []
+    # Perform the "task".
+    for page in environment['pages']:
+        try:
+            r = requests.get("https://" + domain + page, allow_redirects=True, timeout=4)
+            htmlsoup = BeautifulSoup(r.text, 'lxml')
+            # get title and put in dupe-checking list
+            title = htmlsoup.find('title').get_text()
+            titles.append(title)
+            # and description
+            description = htmlsoup.select_one("meta[name='description']")
+            if description:
+                descriptions.append(description['content'])
+            # and can we find dc:date?
+            dc_date = htmlsoup.select_one("meta[name='DC.Date']")
+            # if we found one, grab the content
+            if dc_date:
+                dc_date = dc_date['content']
+
+            # Find the main tag (or alternate)
+            # Potential TO-DO: check that there is only one. Necessary? ¯\_(ツ)_/¯
+            maintag = True if htmlsoup.find('main') else False
+            # if we couldn't find `main` look for the corresponding role
+            if not maintag:
+                maintag = True if htmlsoup.select('[role=main]') else False
+
+            results[page] = {
+                'page': page,
+                'status': str(r.status_code),
+                'title': title,
+                'description': description,
+                'has main tag': maintag,
+                'DC Date': dc_date
+            }
+        except Exception as error:
+            results[page] = "Could not get data from %s%s: %s" % (domain, page, error)
+
+    # now check for dupes
+    if len(titles) != len(set(titles)):
```

Comment:
Thoughts on using `collections.Counter` for the dupe-checking instead? We could avoid creating the lists and the set and just create the `Counter`.

Reply:
No thoughts. Never used it. It _feels_ like overkill given the small size of the lists, but I could easily be persuaded otherwise.

Reply:
I don't feel strongly about this one. I think we'd need two Counters anyway to distinguish `titles` vs `descriptions so I don't think it'd actually clean it up.

Reply:
I also thought about just throwing the error the first time x is in list. Maybe. Again, no strong opinions.